Fix idle priority task during animation

AsturaPhoenix opened the pull request:
This PR is based on https://github.com/flutter/flutter/pull/82781 but takes a slightly different approach where instead of polling the task queue on the `Timer` queue, it waits until after the post-frame callback so that it doesn't compete with animation tasks scheduled via `scheduleFrameCallback`.

This also attempts to address the comments on the test by mimicking the calls at work and cleaning up earlier tests rather than using a real animator, at the expense of possible fidelity. An integration test may be warranted.

https://user-images.githubusercontent.com/13222934/205311407-82fd5676-e8fb-4b9d-b2e0-20b6b6fe929f.mp4

This video shows this patch at work in https://github.com/AsturaPhoenix/trip_planner_aquamarine / http://34.83.198.158/trip_planner_aquamarine. Here, the marker z-ordering is changed depending on which tab is active, and the change is an expensive operation that is scheduled with `Priority.animation - 1`.

**Issues:** https://github.com/flutter/flutter/issues/82016

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

AsturaPhoenix committed:
Test for idle priority task during animation

Adds a test exercising `scheduleTask` with `Priority.idle` while a
`Ticker` is going off. This test demonstrates the issue by failing and
is fixed by the next commit.

Also necessitates cleaning up one of the previous tests to execute
microtasks since warmup frames have an async, and simply running the
exposed handler leaves dirty state.

AsturaPhoenix committed:
Fix idle priority task during animation

Schedules tasks with priority < `Priority.animation` to possibly execute
as a post-frame callback. Tasks with priority >= `Priority.animation`
are executed at that time, and if there are no scheduled transient tasks
awaiting the next frame then tasks with lower priority are executed as
well.

This has also been tested visually in an app.

jonahwilliams commented:
It looks like you may have run `dart format` on the framework code. We don't use machine formatting, there is more information on how we do format in the style guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

Please revert these formatting changes so it is easier to review your patch

AsturaPhoenix committed:
Fix new test interaction under randomized ordering

The scheduler expects raw time to be monotonic, so avoid using time in
the new test since it's not required.

AsturaPhoenix commented:
@jonahwilliams Thanks; done.

Also fixed the test failure under randomized ordering by avoiding incrementing time during the new test. There seems to be a complex interaction in `resetEpoch` that I thought it best not to touch for now.

AsturaPhoenix committed:
Move queue drain in scheduler test to tearDown

This helps ensure that the queues are drained even if the test fails.

goderbauer commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -528,8 +528,14 @@ mixin SchedulerBinding on BindingBase {
       }
       return _taskQueue.isNotEmpty;
+    } else {
+      addPostFrameCallback((Duration timeStamp) {
```
Does this actually fix the problem of tasks not running, though? Adding a postframecallback does not force a frame. So, if currently no frame is running or scheduled, this may never run and tasks will never execute. I believe, there's no guarantee that a frame is running or scheduled at this point?

AsturaPhoenix replied:
@goderbauer For now I've just added a comment to `SchedulingStrategy` specifying the current behavior, that the strategy is re-evaluated after the next (or current) frame, or when a new task is scheduled. I haven't added any new assertions. I think this may be the least impactful way to go.

goderbauer replied:
It is odd to rely on the inner workings of the default scheduling strategy here, because that one could be fully replaced.

I don't have a great immediate idea of what to do here instead, though. Maybe @Hixie has some advice?

AsturaPhoenix replied:
A different way of looking at it with the documentation change is that rather than relying on the inner workings of the default scheduling strategy, we're defining that when a scheduling strategy returns false, tasks won't be re-evaluated again until "after the next (or current) frame, or until a new task is scheduled."

This shifts the implementation detail dependence so that it's contained within the expectations of the scheduling strategy itself, such that this patch would fix the expectation in the context of the default scheduling strategy. Were a scheduling strategy to be defined that returns false even if there are no frames scheduled, it would then not intrinsically have an expectation that the task/frame would be guaranteed to be scheduled, which I think is okay.

Hixie replied:
Right, using `addPostFrameCallback` doesn't work in the case where, say, the scheduling strategy is "do tasks except on Tuesdays". On a Tuesday, the scheduler strategy would return false, but on Wednesday we wouldn't know to retrigger the task queue.

I wonder if anyone has ever used the scheduling strategy logic. If not, we could just remove it and hard-code our current strategy and then the assumption would be safe and `addPostFrameCallback` would work.

I don't think, with the current API, that adding `addPostFrameCallback` is a good solution because it prevents us from feeling the pain of the broken API (without making the API less broken).

AsturaPhoenix committed:
Specify when 'SchedulingStrategy' is reevaluated

Add a comment to `SchedulingStrategy` specifying that skipped tasks are
re-evaluated after the next or current frame, or after a new task is
scheduled.

AsturaPhoenix commented:
There's another thing that warrants discussion. `_ensureEventLoopCallback` schedules the event servicing loop using `Timer.run`. However, in widget tests this can leave us with a somewhat spurious "A Timer is still pending even after the widget tree was disposed." error. (We do need to ensure that the task we scheduled doesn't try to operate on the widget tree while `!mounted`, though.)

```dart
  void _ensureEventLoopCallback() {
    assert(!locked);
    assert(_taskQueue.isNotEmpty);
    if (_hasRequestedAnEventLoopCallback) {
      return;
    }
    _hasRequestedAnEventLoopCallback = true;
    Timer.run(_runTasks);
  }
```

`WidgetTester.pumpAndSettle` gets us past this, but if an animation was indeed scheduled, then we need at least another `await tester.pump(Duration.zero);` afterwards to service the timer, and this precludes the ability to actually write a widget test to catch the case where a task might try to operate on a disposed widget.

A full solution could be to make tasks cancellable and dispose the timer if all pending tasks are cancelled, but that doesn't seem worth the complexity.

We can get around this by scheduling the event loop as a microtask, though this does effectively increase its priority somewhat. I don't think the effect should be otherwise noticeable.

However, I'm inclined to just live with this. Thoughts?

Example of what such a test needs to look like:

```dart
    await tester.pumpWidget(MyWidget());
    await tester.pumpAndSettle();
    await tester.pump(Duration.zero);
    expect(find.byType(MyChild), findsOneWidget);
```

AsturaPhoenix commented:
@Piinks Yes; @goderbauer raised a concern that I think stems from the limitations of the API.

<details><summary>Please be advised that although I have this issue patched on my fork, my project is no longer hitting this issue.</summary>

I was using this to deal with it being expensive to change z-ordering of a large number of markers on Google Maps, but this is no longer an issue now that we've migrated to [`flutter_map`](https://pub.dev/packages/flutter_map). Google Maps seems to be doing a lot of unfortunate serialization, particularly on Android, where it's harder to avoid as it happens partly in GMS.
</details>

AsturaPhoenix commented:
I'm inclined to close this PR since we no longer require this behavior.
